tests/StateMachineTest.php

```diff
@@ -0,0 +1,341 @@
+    }
+
+    public function test_it_should_trigger_an_event_before_any_transition()
+    {
+        $subscriber = new TestSubscriber();
+        $context = TestContext::fromString('started');
+        $machine = StateMachine::create($context)
+            ->addTransition('finish', 'started', 'finished')
+            ->addSubscriber($subscriber)
+        ;
+
+        $this->assertFalse($subscriber->methodWasCalled('beforeTransition'));
+        $machine->transitContext($context, 'finish');
+        $this->assertTrue($subscriber->methodWasCalled('beforeTransition'));
+
+        /**
+         * @var TransitionWasRequested $event
+         */
+        $event = $subscriber->triggeredEvent('beforeTransition');
+        $this->assertInstanceOf(TransitionWasRequested::class, $event);
+        $this->assertInstanceOf(StateTransition::class, $event->transition());
+    }
+
+    public function test_it_should_trigger_an_event_after_any_transition()
+    {
+        $subscriber = new TestSubscriber();
+        $context = TestContext::fromString('started');
+        $machine = StateMachine::create($context)
+            ->addTransition('finish', 'started', 'finished')
+            ->addSubscriber($subscriber)
+        ;
+
+        $this->assertFalse($subscriber->methodWasCalled('afterTransition'));
+        $machine->transitContext($context, 'finish');
+        $this->assertTrue($subscriber->methodWasCalled('afterTransition'));
+
+        /**
+         * @var TransitionWasSuccessful $event
+         */
+        $event = $subscriber->triggeredEvent('afterTransition');
+        $this->assertInstanceOf(TransitionWasSuccessful::class, $event);
+        $this->assertInstanceOf(StateTransition::class, $event->transition());
+    }
+
+    public function test_it_should_trigger_a_custom_event_before_a_specific_transition()
+    {
+        $subscriber = new TestSubscriber();
+        $context = TestContext::fromString('started');
+        $machine = StateMachine::create($context)
+            ->addSubscriber($subscriber)
+            ->addTransition('finish', 'started', 'finished');
+
+        $this->assertFalse($subscriber->methodWasCalled('beforeStartToFinish'));
+        $machine->transitContext($context, 'finish');
+        $this->assertTrue(
+            $subscriber->methodWasCalled('beforeStartToFinish'),
+            'The start to finish transition event should be triggered on before'
+        );
+
+        /**
+         * @var ContextTransitionWasRequested $event
+         */
+        $event = $subscriber->triggeredEvent('beforeStartToFinish');
+        $this->assertInstanceOf(ContextTransitionWasRequested::class, $event);
+        $this->assertSame($context, $event->context());
+    }
+
+    public function test_it_should_trigger_a_custom_event_after_a_specific_transition()
+    {
+        $subscriber = new TestSubscriber();
+        $context = TestContext::fromString('started');
```

Comment:
Use example instead. 

Add `CarWindow` State where the context can affect the transition. (When lock passenger  window may be opened, but not other)